### PR TITLE
Leave `context.channels` empty by default

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

- Closes https://github.com/conda/conda/issues/14178
- Closes https://github.com/conda/conda/issues/12356

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda/base/context.py [hunk 1]
```diff
@@ -520,6 +520,19 @@ def post_build_validation(self):
                 "Only one can be set to 'True'.",
             )
             errors.append(error)
+        if not self.channels:
+            from conda.core.prefix_data import PrefixData
+
+            conda_meta_info = next(PrefixData(self.root_prefix).query("conda"), None)
+            if conda_meta_info:
+                conda_channel = conda_meta_info.channel
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/base/context.py:528]: It's there, but as a URL. I need to parse it into a `Channel`, though, thanks for the tip!

File: conda/base/context.py [hunk 2]
```diff
@@ -520,6 +518,20 @@ def post_build_validation(self):
...
+
+            conda_meta_info = next(PrefixData(self.root_prefix).query("conda"), None)
+            if conda_meta_info:
+                conda_channel = Channel(conda_meta_info.channel).canonical_name
+            else:
+                conda_channel = "[desired channel]"
+            log.warning(
+                "Channel list is empty. Please run 'conda config --add channels %s' to "
+                "remove this warning.",
+                conda_channel,
```
Comment (jezdez) [inline conda/base/context.py:533]: Just note that this'll require more info I think, but the word-smithing isn't important at this stage

File: conda/core/subdir_data.py [hunk 3]
```diff
@@ -113,6 +113,7 @@ def query_all(
         create_cache_dir()
         if channels is None:
             channels = context.channels
+            # TODO: Raise if 'channels' is empty?
```
Comment (jezdez) [inline conda/core/subdir_data.py:116]: Yes, I think that would imply that conda is misconfigured

File: conda/base/context.py [hunk 4]
```diff
@@ -936,23 +948,6 @@ def channels(self):
...
-            # We needs to get the config_files and then check that they
-            # don't define channels
-            channel_in_config_files = any(
-                "channels" in context.raw_data[rc_file].keys()
-                for rc_file in self.config_files
-            )
-            if argparse_channels and not channel_in_config_files:
-                return tuple(
-                    IndexedSet((*local_add, *argparse_channels, DEFAULTS_CHANNEL_NAME))
-                )
```
Comment (jezdez) [inline conda/base/context.py:954]: I'm thinking removing this outright might surprise users, and we probably should raise a [FutureWarning](https://docs.python.org/3/library/exceptions.html#FutureWarning) about it. @kenodegard would you say it would suffice to add a similar `Channel list is empty. Please run 'conda config --add channels %s' to ...` warning here for one or two releases?

Comment (jaimergp) [reply conda/base/context.py:954]: I'll proceed the warning assuming Ken is alright with this direction.

Comment (kenodegard) [reply conda/base/context.py]: @jaimergp you should use `deprecated.topic` for this

File: conda/cli/install.py
Comment (kenodegard) [inline conda/cli/install.py:1]: why this change?

File: conda/testing/notices/fixtures.py [hunk 5]
```diff
@@ -35,6 +36,7 @@ def notices_mock_fetch_get_session():
 @pytest.fixture(scope="function")
 def conda_notices_args_n_parser():
     parser = generate_parser()
-    args = parser.parse_args(["notices"])
+    args = parser.parse_args(["notices", "-c", DEFAULTS_CHANNEL_NAME])
```
Comment (kenodegard) [inline conda/testing/notices/fixtures.py:39]: Using the full option argument makes everything less magical and easier to read:

```suggestion
    args = parser.parse_args(["notices", f"--channel={DEFAULTS_CHANNEL_NAME}"])
```

File: conda/cli/install.py
Comment (jaimergp) [reply conda/cli/install.py:1]: Because `context.channels` is a property, if you use it twice, you'll get the warning twice in the CLI. There's a third occurrence happening in conda-libmamba-solver I'm fixing at https://github.com/conda/conda-libmamba-solver/pull/525.

File: conda/base/context.py (hunk 4 above)
Comment (jaimergp) [reply conda/base/context.py]: I've replaced them accordingly but now I don't get the warnings in the CLI 🤔 

This is what I'm using:

```python
if self._channels:
    _channels = self._channels
else:
    deprecated.topic(
        "24.9",
        "25.1",
        topic=f"Adding '{DEFAULTS_CHANNEL_NAME}' to channel list implicitly.",
        addendum="To remove this warning, please choose a default channel explicitly "
        "via 'conda config --add channels <name>'. In the future, the implicit "
        "default channel configuration will be removed.",
    )
    _channels = [DEFAULTS_CHANNEL_NAME]
```

File: conda/cli/install.py
Comment (kenodegard) [reply conda/cli/install.py:1]: unless the warning filters are overriden the default behavior is to only print the warning the first time it is encountered:

Screenshot 2024-09-12 at 12 14 45

and verified with this sample code:

Screenshot 2024-09-12 at 12 16 56

File: conda/base/context.py (hunk 4 above)
Comment (kenodegard) [reply conda/base/context.py]: by default `deprecated.topic` uses a DeprecationWarning which is ignored by default (since its a developer warning, not a user warning)

you also need to specify `deprecation_type=FutureWarning`:

```python
if self._channels:
    _channels = self._channels
else:
    deprecated.topic(
        "24.9",
        "25.1",
        topic=f"Adding '{DEFAULTS_CHANNEL_NAME}' to channel list implicitly.",
        addendum=(
            "To remove this warning, please choose a default channel explicitly "
            "via 'conda config --add channels <name>'. In the future, the implicit "
            "default channel configuration will be removed."
        ),
        deprecation_type=FutureWarning,
    )
    _channels = [DEFAULTS_CHANNEL_NAME]
```

File: conda/cli/install.py
Comment (jaimergp) [reply conda/cli/install.py:1]: All I can say is that my CLI tests in the devcontainer printed the same warning several times 🤷

File: conda/base/context.py (hunk 4 above)
Comment (jaimergp) [reply conda/base/context.py]: I still don't see it but I think it's due to this clause in the generate_message logic, right?

```
        if self._version_less_than(deprecate_in):
            category = PendingDeprecationWarning
            warning = f"is pending deprecation and will be removed in {remove_in}."
```

Comment (kenodegard) [reply conda/base/context.py]: correct, while we are in the pending deprecation window you will need to explicitly enable all warnings to see the warnings (e.g., `PYTHONWARNINGS=d`), since `conda 24.9` hasn't been tagged yet the deprecation logic still interprets this as being a pending deprecation window

File: conda/base/context.py [hunk 6]
```diff
@@ -949,11 +947,32 @@ def channels(self):
                 for rc_file in self.config_files
             )
             if argparse_channels and not channel_in_config_files:
+                deprecated.topic(
+                    "24.9",
+                    "25.1",
```
Comment (jezdez) [inline conda/base/context.py:952]:
```suggestion
                    "25.3",
```

File: conda/base/context.py [hunk 7]
```diff
@@ -949,11 +947,32 @@ def channels(self):
...
             )
             if argparse_channels and not channel_in_config_files:
+                deprecated.topic(
+                    "24.9",
+                    "25.3",
+                    topic=f"Adding '{DEFAULTS_CHANNEL_NAME}' to channel list implicitly.",
+                    addendum="To remove this warning, please choose a default channel explicitly "
+                    "via 'conda config --add channels <name>'. In the future, the implicit "
+                    "default channel configuration will be removed.",
+                    deprecation_type=FutureWarning,
```
Comment (jezdez) [inline conda/base/context.py:957]: Let's move the addendum into a variable if it's repeated below

File: conda/base/context.py [hunk 8]
```diff
@@ -949,11 +947,32 @@ def channels(self):
...
                     IndexedSet((*local_add, *argparse_channels, DEFAULTS_CHANNEL_NAME))
                 )
-
-        return tuple(IndexedSet((*local_add, *self._channels)))
+        if self._channels:
+            _channels = self._channels
+        else:
+            deprecated.topic(
+                "24.9",
+                "25.1",
```
Comment (jezdez) [inline conda/base/context.py:967]:
```suggestion
                "25.3",
```

**Review by jezdez — CHANGES_REQUESTED** [review]

File: conda/base/context.py [hunk 9]
```diff
@@ -949,11 +947,32 @@ def channels(self):
...
             )
             if argparse_channels and not channel_in_config_files:
+                deprecated.topic(
+                    "24.9",
+                    "25.1",
+                    topic=f"Adding '{DEFAULTS_CHANNEL_NAME}' to channel list implicitly.",
+                    addendum="To remove this warning, please choose a default channel explicitly "
+                    "via 'conda config --add channels <name>'. In the future, the implicit "
+                    "default channel configuration will be removed.",
+                    deprecation_type=FutureWarning,
```
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda/base/context.py:957]: Since we want the **user** to see this we need `FutureWarning`.

File: conda/base/context.py (hunk 7 above)
Comment (kenodegard) [reply conda/base/context.py:957]: how about we create a new property instead (and mark it for immediate deprecation) to solve this:

```python
    @property
    @deprecated(
        "24.9",
        "25.3",
        addendum=(
            "To remove this warning, please choose a default channel explicitly "
            "via 'conda config --add channels <name>'. In the future, the implicit "
            "default channel configuration will be removed."
        ),
        deprecation_type=FutureWarning,
    )
    def defaults(self) -> str:
        return DEFAULTS_CHANNEL_NAME
```

use this property instead of the `DEFAULTS_CHANNEL_NAME` within the `channels` property

File: conda/cli/main_config.py [hunk 10]
```diff
@@ -731,8 +731,6 @@ def execute_config(args, parser):
     for arg, prepend in zip((args.prepend, args.append), (True, False)):
         for key, item in arg:
             key, subkey = key.split(".", 1) if "." in key else (key, None)
-            if key == "channels" and key not in rc_config:
-                rc_config[key] = ["defaults"]
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/cli/main_config.py]: Fair enough, but that means we'll have to do something in setup-miniconda to prevent the implicit `defaults` addition. Either the `nodefaults` hack or a boolean to implement this explicit behavior.

File: conda/core/subdir_data.py (hunk 3 above)
Comment (jezdez) [reply conda/core/subdir_data.py:116]: Let's file a separate ticket about this?

File: news/14227-nodefaults [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+### Deprecations
+
+* The `defaults` multichannel will stop being the (implicit) default value for `channels`. Users
+  relying on this behaviour are encouraged to run `conda config --add channels defaults`.
+  This is now deprecated, and will become an error in 25.3. (#14178 via #14227)
+* `conda config --add/--append channels ...` will warn when `defaults` is being added implicitly.
+  By 25.3, this behaviour will be removed and users should run `conda config --add/--append channels defaults` explicitly if needed. (#12356 via #14227)
+* Without an explicit `channels` configuration (`condarc`, environment variables, CLI flags),
+  `context.channels` will warn about emitting `defaults` implicitly. In 25.3, an empty list will be
+  returned. If this behavior is desired, the `context.channels or [Channel("defaults")]` idiom is recommended. (#14227)
```
Comment (jezdez) [inline news/14227-nodefaults:18]:
```suggestion
* Without an explicit `channels` configuration (via `condarc` files, environment
  variables, or CLI flags), conda will warn about using `defaults` implicitly.
  In 25.3, an empty list will be used. (#14227)
```

File: news/14227-nodefaults [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+* <news item>
+
+### Deprecations
+
+* The `defaults` multichannel will stop being the (implicit) default value for `channels`. Users
+  relying on this behaviour are encouraged to run `conda config --add channels defaults`.
+  This is now deprecated, and will become an error in 25.3. (#14178 via #14227)
+* `conda config --add/--append channels ...` will warn when `defaults` is being added implicitly.
+  By 25.3, this behaviour will be removed and users should run `conda config --add/--append channels defaults` explicitly if needed. (#12356 via #14227)
```
Comment (jezdez) [inline news/14227-nodefaults:15]:
```suggestion
* Running `conda config --add/--append channels ...` will warn when `defaults` is being
  added implicitly. By 25.3, this behaviour will be removed and users should run
  `conda config --add/--append channels defaults` explicitly if needed. Conda
  distribution installers like miniforge or miniconda will pre-
  configure conda channels during installation. (#12356 via #14227)
```

File: news/14227-nodefaults [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+### Bug fixes
+
+* <news item>
+
+### Deprecations
+
+* The `defaults` multichannel will stop being the (implicit) default value for `channels`. Users
+  relying on this behaviour are encouraged to run `conda config --add channels defaults`.
+  This is now deprecated, and will become an error in 25.3. (#14178 via #14227)
```
Comment (jezdez) [inline news/14227-nodefaults:13]:
```suggestion
* The `defaults` multichannel will stop being the (implicit) default value for `channels`. Users
  relying on this behaviour are encouraged to run `conda config --add channels defaults`.
  This is now pending deprecation, and will be fully deprecated in 25.3. (#14178 via #14227)
```

**Review by kenodegard — APPROVED** [review]

File: conda/cli/install.py
Comment (jezdez) [reply conda/cli/install.py:1]: That's odd though, I haven't checked with the devcontainer, does this happen locally for you as well @jaimergp?

**Review by jezdez — APPROVED** [review]
